dronegen/tag.go

```diff
@@ -0,0 +1,425 @@
+func tagPipelines() []pipeline {
+	var ps []pipeline
+	// regular tarball builds
+	for _, arch := range []string{"amd64", "386", "arm", "arm64"} {
+		for _, fips := range []bool{false, true} {
+			if (arch == "386" || arch == "arm") && fips {
+				// FIPS mode not supported on i386 or ARM
+				continue
+			}
+			ps = append(ps, tagPipeline(tagBuildType{os: "linux", arch: arch, fips: fips}))
+			// TODO(gus): support needs to be added upstream for building ARM/ARM64 packages first
+			// ps = append(ps, tagPackagePipeline(rpmPackage, tagBuildType{os: "linux", arch: arch, fips: fips}))
+			// ps = append(ps, tagPackagePipeline(debPackage, tagBuildType{os: "linux", arch: arch, fips: fips}))
+		}
+	}
+
+	// TODO(gus): needed until support is added upstream for building ARM/ARM64 packages
+	// Remove this section and uncomment above once this is added.
+	for _, arch := range []string{"amd64", "386"} {
+		for _, fips := range []bool{false, true} {
+			if (arch == "386" || arch == "arm") && fips {
+				// FIPS mode not supported on i386 or ARM
+				continue
+			}
+			ps = append(ps, tagPackagePipeline(rpmPackage, tagBuildType{os: "linux", arch: arch, fips: fips}))
+			ps = append(ps, tagPackagePipeline(debPackage, tagBuildType{os: "linux", arch: arch, fips: fips}))
+		}
+	}
+
+	// Only amd64 Windows is supported for now.
+	ps = append(ps, tagPipeline(tagBuildType{os: "windows", arch: "amd64"}))
+	// Also add the two CentOS 6 artifacts.
+	ps = append(ps, tagPipeline(tagBuildType{os: "linux", arch: "amd64", centos6: true}))
+	ps = append(ps, tagPipeline(tagBuildType{os: "linux", arch: "amd64", centos6: true, fips: true}))
+	return ps
+}
+
+// tagPipeline generates a tag pipeline for a given combination of os/arch/FIPS
+func tagPipeline(params tagBuildType) pipeline {
+	if params.os == "" {
+		panic("params.os must be set")
+	}
+	if params.arch == "" {
+		panic("params.arch must be set")
+	}
+
+	pipelineName := fmt.Sprintf("build-%s-%s", params.os, params.arch)
+	pipelineTitleSegment := "release artifacts"
+	if params.centos6 {
+		pipelineName += "-centos6"
+		pipelineTitleSegment = "CentOS 6 release artifacts"
+	}
+	tagEnvironment := map[string]value{
+		"UID":    value{raw: "1000"},
+		"GID":    value{raw: "1000"},
+		"GOPATH": value{raw: "/go"},
+		"OS":     value{raw: params.os},
+		"ARCH":   value{raw: params.arch},
+	}
+	if params.fips {
+		pipelineName += "-fips"
+		tagEnvironment["FIPS"] = value{raw: "yes"}
+		if params.centos6 {
+			pipelineTitleSegment = "CentOS 6 FIPS release artifacts"
+		} else {
+			pipelineTitleSegment = "FIPS release artifacts"
+		}
+	}
+
+	p := newKubePipeline(pipelineName)
+	p.Environment = map[string]value{
+		"RUNTIME": value{raw: "go1.15.5"},
+	}
+	p.Trigger = triggerTag
+	p.Workspace = workspace{Path: "/go"}
+	p.Volumes = []volume{
+		volumeDocker,
+	}
+	p.Services = []service{
+		{
+			Name:  "Start Docker",
+			Image: "docker:dind",
+			Volumes: []volumeRef{
+				volumeRefDocker,
+			},
+		},
```

Comment:
should this be in `common.go` and used by other pipelines?

Reply:
I think I was avoiding this so far because `services` and `volumes` are tightly coupled which means there's a minimum of two common structs and it didn't seem entirely worth it.

There are also three slightly different versions of this used:
- one for regular `docker:dind` with `/var/run` (most common)
- one with the above plus `/tmpfs` for in-memory operations like lint/unit/integration tests
- one with both of the above plus `/tmp` for a second `emptyDir` during integration tests

I'll see if I can make the common use case simpler.